browser/brave_ads/delegate/ads_service_delegate_impl.cc

```diff
@@ -54,16 +68,49 @@ mojo::Remote<bat_ads::mojom::BatAdsService> LaunchOutOfProcessBatAdsService() {
 
 }  // namespace
 
-BatAdsServiceFactoryImpl::BatAdsServiceFactoryImpl() = default;
+AdsServiceDelegateImpl::AdsServiceDelegateImpl(Profile* profile)
+    : profile_(profile) {
+  CHECK(profile_);
+}
 
-BatAdsServiceFactoryImpl::~BatAdsServiceFactoryImpl() = default;
+AdsServiceDelegateImpl::~AdsServiceDelegateImpl() = default;
 
-mojo::Remote<bat_ads::mojom::BatAdsService> BatAdsServiceFactoryImpl::Launch()
-    const {
+mojo::Remote<bat_ads::mojom::BatAdsService>
+AdsServiceDelegateImpl::LaunchBatAdsService() const {
   DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
 
   return ShouldLaunchAsInProcessService() ? LaunchInProcessBatAdsService()
                                           : LaunchOutOfProcessBatAdsService();
 }
 
+void AdsServiceDelegateImpl::OpenNewTabWithUrl(const GURL& url) {
+  if (g_browser_process->IsShuttingDown()) {
+    return;
+  }
+
+  if (!url.is_valid()) {
+    return VLOG(1) << "Failed to open new tab due to invalid URL: " << url;
+  }
+
+#if BUILDFLAG(IS_ANDROID)
+  // ServiceTabLauncher can currently only launch new tabs
+  const content::OpenURLParams params(url, content::Referrer(),
+                                      WindowOpenDisposition::NEW_FOREGROUND_TAB,
+                                      ui::PAGE_TRANSITION_LINK, true);
+  ServiceTabLauncher::GetInstance()->LaunchTab(
+      profile_, params, base::BindOnce([](content::WebContents*) {}));
+#else
+  Browser* browser = chrome::FindTabbedBrowser(profile_, false);
+  if (!browser) {
+    browser = Browser::Create(Browser::CreateParams(profile_, true));
+  }
+
+  NavigateParams nav_params(browser, url, ui::PAGE_TRANSITION_LINK);
+  nav_params.disposition = WindowOpenDisposition::SINGLETON_TAB;
+  nav_params.window_action = NavigateParams::SHOW_WINDOW;
+  nav_params.path_behavior = NavigateParams::RESPECT;
+  Navigate(&nav_params);
+#endif
```

Comment:
How about we decouple Android and other implementation.

Reply:
Agreed to do it as a follow-up. In the follow-up PR this presubmit warning should also be fixed:

```
Banned functions were used.
    browser/brave_ads/delegate/ads_service_delegate_impl.cc:103:
      Most code should already be scoped to a Browser. Pass in a Browser* or retreive from an existing entity with a reference to the Browser.
```

Reply:
Thank you